src/gordon_gcp/clients/http.py

```diff
@@ -95,15 +92,96 @@ def __init__(self, auth_client=None, session=None):
             body (obj): (optional) A dictionary, bytes, or file-like
                 object to send in the body of the request.
             headers (dict): (optional) HTTP headers to send with the
-                request. Headers pass through to the request will
+                request. Headers passed through to the request will
                 include :attr:`DEFAULT_REQUEST_HEADERS`.
+            retry_type (str): (optional) The type of retries to use: "none",
+                "simple", (retry up to tries/timeout), "no4xx" (don't retry
+                on 4xx, otherwise up to tries/timeout), or "custom".  If
+                "custom", uses the retry predicate specified in the other
+                parameters.  See also extra_401_retry.
+            tries (int): (optional) Total maximum number of times to try the
+                request, including the first (i.e. 1 == no retries).  The number
+                of actual tries is subject to retry_type and timeout as well as
+                this value.
+            timeout (int): (optional) Total maximum seconds to keep trying the
+                request (None == unlimited).  The number of actual tries is
+                subject to retry_type and tries as well as this value.
+            retry_predicate (callable): (optional) Function to call to decide
+                if we should keep trying the request.  Ignored unless
+                retry_type is "custom".  Must return True for "continue retrying
+                (subject to tries and timeout)" or False for "stop retrying".
+                Last argument must be the response from _request().  If the
+                function takes any other arguments, their values must be
+                supplied in predicate_args and/or predicate_kwargs.
+            predicate_args (list): (optional) See retry_predicate.
+            predicate_kwargs (dict): (optional) See retry_predicate.
+            extra_401_retry (bool): (optional) If True, run one more request
+                attempt on a 401 (unauthorized), regardless of the value of
+                retry_type.  This makes sure we've tried generating a new token.
         Returns:
             (str) HTTP response body.
         Raises:
-            :exc:`.GCPHTTPError`: if any exception occurred.
+            :exc:`.GCPHTTPError`: If a response code >=400 was received and
+                there are no more tries (or no retrying is specified).
         """
-        refresh_attempt = kwargs.pop('cred_refresh_attempt', 0)
+        retry_predicates = {
+            "none": lambda resp: False,  # never retry
```

Comment:
If we end up going down this route, I think it's more pythonic to do something like this:

```py
# retry_type=None in method kwargs
predicate = lambda resp: False  # set default
retry_predicates = {
    'all': lambda resp: resp.status >= 400,
    'server_errors': lambda resp: resp.status >= 500,
    'custom': retry_predicate
}
try:
    predicate = retry_predicate[retry_type]
except KeyError:
    pass  # end up using the default
```

Reply:
That would silently swallow a bad value.  I wanted to cause a KeyError on bad value.

Reply:
I can note that explicitly if that changes anything.